client/core/core.go

```diff
@@ -5118,6 +5111,48 @@ func processPreimageRequest(c *Core, dc *dexConnection, reqID uint64, oid order.
 	return nil
 }
 
+// acceptCsum will record the commitment checksum so we can verify that the
+// subsequent match_proof with this order has the same checksum. If it does not,
+// the server may have used the knowledge of this preimage we are sending them
+// now to alter the epoch shuffle.
+func acceptCsum(dc *dexConnection, reqID uint64, tracker *trackedTrade, isCancel bool, commitChecksum dex.Bytes) error {
+	var respondWithErr bool
+
+	// Allow to initialize csum only once per order to prevent possible
+	// malicious behavior.
+	tracker.mtx.Lock()
+	if isCancel {
+		if tracker.cancel.csum != nil {
+			respondWithErr = true
+		} else {
+			tracker.cancel.csum = commitChecksum
+		}
+	} else {
+		if tracker.csum != nil {
```

Comment:
Checking `!= nil` is one possibility I suggested in the previous PR discussion.  Another is non-nil *and* `!bytes.Equal`, which would allow client to respond with the order's preimage as long as the server didn't change up the csum from the initial request.

While non-nil is a good indicator that server has already asked and something is fishy, I can see the latter being a legit case if server has to retry the request because of perhaps a ill-timed restart or even a timeout on a previously request because of a broken connection (note that the csum is stored by the client in tracker *before* attempting to send the `PreimageResponse` and that `Send` could fail - server is not explicitly prohibited from asking again).

Essentially, if the csum field is already set, we clearly don't need to rewrite it, but we don't have to reject/ignore the requests unless the csum would change value.

Reply:
Makes sense, will implement this soon.

Reply:
Done.